woocommerce/woocommerce
# WC_Product_Variable_Data_Store_CPT: Add variable product transient validation

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:



For a summary of the bug this fixes please see here https://github.com/woocommerce/woocommerce/issues/50855#issuecomment-2531651353

We are validating variable product transient data before using it. This is because there is an ongoing and widespread issue whereby corrupted transient data is being used, this results in products being displayed as out of stock on the frontend until the user updates the product to refresh the transient data.

We don't know what or how this data is being corrupted yet, but in the meantime this PR adds validation so we can avoid using it unless we are certain its valid.



In this PR I add validation to both `wc_product_children_` and `wc_var_prices_` transient data to check their validity.

`validate_children_data`:
* Data is an array
* Has required keys 'all' and 'visible'
* Both 'all' and 'visible' are arrays
* If version exists, matches current version
* All IDs in both arrays are numeric values

`validate_prices_data`
* Data is an array
* If version exists, matches current version
* Has required price type arrays: 'price', 'regular_price', 'sale_price'
* All price type entries are arrays
* Variation IDs are numeric
* Price values are either numeric or empty strings
* Each variation ID has entries for all price types

We have added a new `version` property to the transient so we check this exists before validating it to ensure backwards compat.

Closes https://github.com/woocommerce/woocommerce/issues/50855



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Reproducing this issue reliably has proven difficult. As a result I think its best to do a full regression test of adding, viewing and purchasing variable products to ensure no regressions or errors occur. On top of that please:

1. Install Transient Manager plugin and do a search for a variable product prices `_transient_wc_var_prices_`
2. If you want to, you can add some `var_dump()` logs for valid/invalid to observe on the frontend near the top of `read_price_data` method
3. This should evaluate valid currently, now update the transient data with the below (invalid code).
4. Refresh the variable product page of the product who's transient you updated, it should evaluate as invalid but should also correctly display the price (because its updated the transient cache after detecting invalid data)
5. Refresh the page again, and it should evaluate as valid (because it was updated in step 4).
6. Do the same with `wc_product_children_` transient data and the `read_children` method.
7. Do some regression testing across classic & block themes archive/category pages and product pages.

```
{
  "version": "1725527855",
  "f9e544f77b7eac7add281ef28ca5559f": {
    "price": [],
    "regular_price": [],
    "sale_price": []
  },
  "f74c5567e2b347f8728ea07774b6f7e4": {
    "price": [],
    "regular_price": [],
    "sale_price": []
  }
}
```



### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
